src/client/storage/index.js

```diff
@@ -68,32 +77,56 @@ const getChunk = async (chunkId, encoding = 'utf8', useCache = true) => {
         log.debug({chunkId}, 'Graphql request success');
 
         for (const edge of queryResult.transactions.edges) {
-            const txid = edge.node.id;
-            log.debug({chunkId, txid}, 'Downloading data from arweave');
+            try {
+                const txid = edge.node.id;
+                log.debug({chunkId, txid}, 'Downloading data from arweave');
+
+                const data = await storage.getDataByTxId(txid);
+                log.debug({chunkId, txid}, 'Successfully downloaded data from arweave');
+
+                const buf = Buffer.from(data);
+
+                const hash = hashFn(buf).toString('hex');
+                if (hash !== chunk.id) {
+                    log.warn(
+                        {chunkId, hash, query, buf: buf.toString()},
+                        'Chunk id and data do not match'
+                    );
+                    continue;
+                }
+
+                await fs.writeFile(chunkPath, buf);
+
+                chunk.size = buf.length;
+                chunk.dl_status = CHUNK_DOWNLOAD_STATUS.COMPLETED;
+                await chunk.save();
+                return buf;
+            } catch (err) {
+                log.warn({chunkId, err}, 'Trying to download chunk from Arweave threw an error');
+            }
+        }
 
-            const data = await storage.getDataByTxId(txid);
-            log.debug({chunkId, txid}, 'Successfully downloaded data from arweave');
+        log.warn(
+            {chunkId},
+            'Chunk not found in Arweave. Falling back to requesting it from bundler backup'
+        );
 
+        // TODO: refactor before mainnet!!!
+        // If we failed to retrieve the chunk from Arweave,
+        // retrieve it from the bundler's backup (S3).
+        try {
+            const url = `${config.get('storage.arweave_bundler_url')}/download/${chunkId}`;
```

Comment:
I'd move this up to constants same as other config values

Reply:
I'll move it.